vochain/state_snapshot.go

```diff
@@ -0,0 +1,422 @@
+package vochain
+
+import (
+	"bytes"
+	"encoding/binary"
+	"encoding/gob"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/vocdoni/arbo"
+	"go.vocdoni.io/dvote/log"
+	"go.vocdoni.io/dvote/statedb"
+)
+
+const (
+	snapshotHeaderVersion = 1
+	snapshotHeaderLenSize = 32
+)
+
+// ErrTreeNotYetFullyRead is returned if FetchNextTree is called before the current
+// tree is fully read.
+var ErrTreeNotYetFullyRead = fmt.Errorf("current tree is not yet fully read")
+
+// A StateSnapshot is a copy in a specific point in time of the blockchain state.
+// The state is supposed to be a list of nested merkle trees.
+// The StateSnapshot contains the methods for building a single file snapshot of
+// the state containing multiple trees.
+// The implementation allows the encoding and decoding of snapshots.
+//
+// The structure of the snapshot encoded file is:
+//
+//  [headerLen][header][tree1][tree2][treeN]
+//
+// - headerlen is a fixed 32 bytes little endian number indicating the size of the header.
+//
+// - header is the Gob encoded structure containing the information of the trees (size, name, etc.).
+//
+// - treeN is the raw bytes dump of all trees.
+type StateSnapshot struct {
+	path               string
+	file               *os.File
+	lock               sync.Mutex
+	header             SnapshotHeader
+	headerSize         uint32
+	currentTree        int // indicates the current tree
+	treeReadPointer    int // points to the last read byte
+	currentTreeReadEnd int // points to the last byte of the current tree
```

Comment:
I'd probably use sized integers here just in case, since this does encoding/decoding. you really don't want to run into issues on 32 bit if e.g. the files near gigabytes in size.

Reply:
No, this vars are not encoded, only SnapshotHeader and SnapshotHeaderTree are and both use sized types.

Reply:
These are counters and its more convenient to use INT since they are used for indexing arrays